pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -376,6 +380,19 @@ func (c *Controller) onServiceEvent(pre, curr *v1.Service, event model.Event) er
 	// Create the standard (cluster.local) service.
 	svcConv := kube.ConvertService(*curr, c.opts.DomainSuffix, c.Cluster())
 
+	preNotVisible := isServiceNotVisible(pre)
+	currNotVisible := isServiceNotVisible(curr)
+	// If the service is consistently not visible, not process it.
+	// If the service changes from being visible to not visible, treat it as a delete event.
+	// If the service changes from being not visible to visible, treat it as an add event.
+	if preNotVisible && currNotVisible {
+		return nil
+	}
+	if currNotVisible {
+		c.deleteService(svcConv)
```

Comment:
not comfortable with deleting it, why do we need this ?

Reply:
This should depend on whether to consider a completely invisible service as non-existent.

Reply:
I don't think it is, and I **strongly** suspect this will break something. All the tests pass, so its clearly not something tested, but we also have no test coverage of `exportTo=~` so that is expected. Even if you added tests, though, I would still have low confidence.